File: Problems/Algorithms/1608. Special Array with X Elements Greater Than Or Equal X/SpecialArray.kt
class Solution {
    fun specialArray(nums: IntArray): Int {
        nums.sortDescending()
        val n = nums.size
        
        var left = 0
        var right = n
        while (left < right) {
            val mid = left + (right - left) / 2
            if (nums[mid] > mid) {
                left = mid + 1
            } else {
                right = mid
            }
        }
        
        return if (left < n && nums[left] == left) return -1 else left
    }
}

File: Problems/Algorithms/1608. Special Array with X Elements Greater Than Or Equal X/special_array.py
#!/usr/bin/env python3

from typing import List


class Solution:
    def specialArray(self, nums: List[int]) -> int:
        n = len(nums)
        
        freq = [0] * (n + 1)
        for num in nums:
            freq[min(n, num)] += 1
        
        num_greater_than_or_equal = 0
        for i in range(n, 0, -1):
            num_greater_than_or_equal += freq[i]
            if i == num_greater_than_or_equal:
                return i
        
        return -1
